File: Hard/42. Trapping Rain Water/solution.java
class Solution {
    public int trap(int[] height) {
        int len = height.length;
        int[] leftToRight = new int[len];
        int[] rightToLeft = new int[len];
        int val = 0;
        for (int i = 0; i < len; i++) {
            val = Math.max(val, height[i]);
            leftToRight[i] = val;
        }
        val = 0;
        for (int i = len - 1; i >= 0; i--) {
            val = Math.max(val, height[i]);
            rightToLeft[i] = val;
        }
        int[] diff = new int[len];
        for (int i = 0; i < len; i++) {
            diff[i] = Math.min(leftToRight[i], rightToLeft[i]);
        }
        int ans = 0;
        for (int i = 0; i < len; i++) {
            ans += diff[i] - height[i];
        }
        return ans;
    }
}

File: Hard/42. Trapping Rain Water/solution.py
'''
    l[]: precompute the max of height[0 -> i - 1]
    r[]: precompute the max of height [i - > n - 1]
    then iterate through these 2 lists, find the max of left part of current index
    and find the max of right part of current index
    time complexity: O(n)
    space complexity: O(n)
'''
class Solution:
    def trap(self, height: List[int]) -> int:
        
        l = [0] * len(height)
        r = [0] * len(height)
        
        for i in range(len(height)):
            if i == 0:
                l[i] = height[i]
            else:
                l[i] = max(height[i], l[i - 1])
            
        
        for i in range(len(height) - 1, -1, -1):
            if i == len(height) - 1:
                r[i] = height[-1]
            else:
                r[i] = max(r[i + 1], height[i])
        answer = 0
        for i in range(len(height)):
            answer += min(r[i], l[i]) - height[i]
        
        return answer